File: backend/app/utils/text_sanitize.py
"""文本清理 — 去除无效 Unicode 代理对，避免 Pydantic/JSON 编码失败。"""

import re


def sanitize_unicode(value: str | None) -> str:
    """移除 lone surrogate，保证可 UTF-8 编码。"""
    if not value:
        return ""
    cleaned = "".join(ch for ch in value if not (0xD800 <= ord(ch) <= 0xDFFF))
    return cleaned.strip()


def normalize_project_name(value: str | None, *, max_len: int = 200) -> str:
    """从可能含 Markdown 的文本中提取简短项目名。"""
    text = sanitize_unicode(value)
    if not text:
        return "未命名项目"

    book = re.search(r"[《「【]([^》」】]{1,40})[》」】]", text)
    if book:
        name = book.group(1).strip()
        if name:
            return name[:max_len]

    heading = re.search(r"^#{1,3}\s*(.+?)(?:\r?\n|$)", text, re.MULTILINE)
    if heading:
        name = heading.group(1).strip()
        name = re.sub(r"创作方案$", "", name).strip(" 《》#")
        if name:
            return name[:max_len]

    text = re.sub(
        r"^(帮我|给我|请|麻烦)\s*(做|生成|创作|制作|写|设计)\s*(一个|一部)?\s*",
        "",
        text,
    ).strip()
    if text:
        first_line = text.splitlines()[0].strip()
        return first_line[:max_len] if first_line else text[:max_len]

    return "未命名项目"

File: app/src/lib/text-sanitize.ts
/** 移除 lone surrogate，避免 JSON / API 编码失败 */
export function sanitizeSurrogates(value: string): string {
  return value.replace(/[\uD800-\uDFFF]/g, '');
}

/** 从创作方案 Markdown 或用户输入提取简短项目名 */
export function normalizeProjectName(value: string, maxLen = 50): string {
  const text = sanitizeSurrogates(value).trim();
  if (!text) return '创作项目';

  const bookMatch = text.match(/[《「【]([^》」】]{1,40})[》」】]/);
  if (bookMatch?.[1]?.trim()) {
    return bookMatch[1].trim().slice(0, maxLen);
  }

  const headingMatch = text.match(/^#{1,3}\s*(.+?)(?:\n|$)/m);
  if (headingMatch?.[1]) {
    const heading = headingMatch[1]
      .replace(/创作方案$/u, '')
      .replace(/[《》]/g, '')
      .trim();
    if (heading) return heading.slice(0, maxLen);
  }

  const cleaned = text
    .replace(/^(帮我|给我|请|麻烦)\s*(做|生成|创作|制作|写|设计)\s*(一个|一部)?\s*/u, '')
    .trim();

  if (cleaned) {
    const firstLine = cleaned.split('\n')[0]?.trim() || cleaned;
    return firstLine.slice(0, maxLen);
  }

  return text.slice(0, maxLen) || '创作项目';
}
